### ISSUE TITLE
[BC Break] Practical design: Concrete classes and benchmarking

### ISSUE THREAD

**Issue by gabbydgab** [php-refactoring/zend-array-utility#2]
Currently, [ArrayUtils](https://github.com/php-refactoring/zend-array-utility/blob/master/src/ArrayUtils.php) class has many responsibility, thus breaking it into specific usage will open more opportunities to enhance its features - also to properly benchmark edge cases of each functionality (e.g. [caching validation](https://github.com/php-refactoring/zend-array-utility/issues) and [array merging strategies](https://github.com/php-refactoring/zend-array-utility/blob/master/benchmark/ArrayMergeBench.php))

### Responsibilities
1. Validation
    - hasStringKey
    - hasIntegerKey
    - hasNumericKeys
    - isList
    - isHasTable
    - isCachable (proposed/new feature) 
2. Merging Strategies
    - Integer and String Keys
    - Integer and String With Preserved Numeric Keys
    - Recursive Merging
    - Merging Null arrays
    - Replacing Keys
    - Removing Keys
    - Prioritized Merging (proposed/new feature)
3. Filtering Strategies
    - Without key filtering
    - Key filtering specific
    - With and without key filtering
4. Iterator to Array Conversion Strategies
    - SimpleIterator
    - RecursiveIterator

### Benchmarking

**Benchmarking the merging  and caching strategies is a MUST**.

One way of validating performance issue is to benchmark, especially, configuration for [Zend\ServiceManager](https://docs.zendframework.com/zend-servicemanager/configuring-the-service-manager/), [Zend\View](https://docs.zendframework.com/zend-view/quick-start/#configuration), [Zend\Router](https://docs.zendframework.com/zend-router/routing/#http-routing-examples).

**Comment by gabbydgab** [comment]
@Ocramius @weierophinney Thoughts?

**Comment by Ocramius** [comment]
@gabbydgab stop any dev work: if this is a split, just make it a split and stop there, with 0 API and functional changes. Further dev to be done once it is under the organisation again.

**Comment by gabbydgab** [comment]
@Ocramius yes, already split the library and tagged it as version 0.1.x (same with master) - including the initial benchmarks (version 0.1.1). With regards to benchmarks, I would like to update them using a practical configuration - larger from what is being tested. Probably will add it as version 0.1.2.

I'm for @weierophinney approval to transfer this repository back to the organization.

I'll be adding more issues that might be needed, like sorting, search and merging by priority.

**Comment by Ocramius** [comment]
Then please wait until it's moved back: one thing at a time, or else this
will just become a mess, and we'll have to start over.
